clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -1254,28 +1274,68 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_switch_sequences()
 
 template <typename RecordType, typename ReaderType>
 typename scheduler_tmpl_t<RecordType, ReaderType>::scheduler_status_t
-scheduler_tmpl_t<RecordType, ReaderType>::get_initial_timestamps()
-{
+scheduler_tmpl_t<RecordType, ReaderType>::get_initial_input_content(
+    bool gather_timestamps)
+{
+    // For every mode, read ahead until we see a filetype record so the user can
+    // examine it prior to retrieving any records.
+    VPRINT(this, 1, "Reading headers from inputs to find filetypes%s\n",
+           gather_timestamps ? " and timestamps" : "");
+    assert(options_.read_inputs_in_init);
     // Read ahead in each input until we find a timestamp record.
     // Queue up any skipped records to ensure we present them to the
     // output stream(s).
     for (size_t i = 0; i < inputs_.size(); ++i) {
         input_info_t &input = inputs_[i];
-        if (input.next_timestamp <= 0) {
+        bool found_filetype = false;
+        bool found_timestamp = !gather_timestamps || input.next_timestamp > 0;
+        if (!found_filetype || !found_timestamp) {
```

Comment:
Some narration would be nice.

// Inputs may have queued records which should be checked first.

// Now read past the queued records until we find the timestamp and filetype.

Better: input_info_t should have an API to get the next record which automatically returns the queued entry if there is one. Would avoid the duplication of this logic.

Reply:
Added comments.

Started trying to write a helper function: but it's actually not simple as we don't want to consume the records in the queue, just observe them, while if we iterate over the real reader we have to push onto the queue.  Hmm.  I think at this point I will add a XXX comment about trying to come up with an improvement to the current code.